src/ralph/backends/database/async_es.py

```diff
@@ -0,0 +1,251 @@
+from ralph.exceptions import BackendException, BackendParameterException
+
+from ..database.base import (
+    BaseAsyncDatabase,
+    DatabaseStatus,
+    StatementParameters,
+    StatementQueryResult,
+    enforce_query_checks,
+)
+from ..database.es import ESQuery, OpType
+
+async_es_settings = settings.BACKENDS.DATABASE.ASYNC_ES
+logger = logging.getLogger(__name__)
+
+
+class AsyncESDatabase(BaseAsyncDatabase):
+    """AsyncElasticsearch database backend."""
+
+    name = "async_es"
+    query_model = ESQuery
+
+    def __init__(
+        self,
+        hosts: list = async_es_settings.HOSTS,
+        index: str = async_es_settings.INDEX,
+        client_options: ESClientOptions = async_es_settings.CLIENT_OPTIONS,
+        op_type: str = async_es_settings.OP_TYPE,
+    ):
+        """Instantiates the AsyncElasticsearch client.
+
+        Args:
+            hosts (list): List of Elasticsearch nodes we should connect to.
+            index (str): The Elasticsearch index name.
+            client_options (dict): A dictionary of valid options for the
+                Elasticsearch class initialization.
+            op_type (str): The Elasticsearch operation type for every document sent to
+                Elasticsearch (should be one of: index, create, delete, update).
+        """
+        self._hosts = hosts
+        self.index = index
+
+        self.client = AsyncElasticsearch(
+            self._hosts, **client_options.dict(), request_timeout=30
+        )
+        if op_type not in [op.value for op in OpType]:
+            raise BackendParameterException(
+                f"{op_type} is not an allowed operation type"
+            )
+        self.op_type = op_type
+
+    async def status(self) -> DatabaseStatus:
+        """Checks Elasticsearch cluster (connection) status."""
+        # Check ES cluster connection
+        try:
+            await self.client.info()
+        except ESConnectionError:
+            return DatabaseStatus.AWAY
+
+        # Check cluster status
+        if "green" not in await CatClient(self.client).health():
+            return DatabaseStatus.ERROR
+
+        return DatabaseStatus.OK
+
+    @enforce_query_checks
+    async def get(self, query: ESQuery = None, chunk_size: int = 500):
+        """Gets index documents and yields them.
+
+        The `query` dictionary should only contains kwargs compatible with the
+        elasticsearch.helpers.async_scan function signature (API reference
+        documentation:
+        https://elasticsearch-py.readthedocs.io/en/latest/async.html#scan).
+        """
+        async for document in async_scan(
+            self.client, index=self.index, size=chunk_size, **query.dict()
+        ):
+            yield document
+
+    async def to_documents(
+        self, stream: TextIO, get_id: Callable[[dict], str]
+    ) -> Generator[dict, None, None]:
+        """Converts `stream` lines to ES documents."""
+        for line in stream:
```

Comment:
```suggestion
        for line in stream:
            await asyncio.sleep(0)
```
An async method that does not await anything seems equivalent to a sync method.
Having a second look at this method, we might benefit from async execution even if no IO operation is done inside it if doing one loop iteration is time-consuming, and other functions perform IO tasks in parallel.
Source: https://stackoverflow.com/questions/65973786/why-does-asyncio-sleep0-make-my-code-faster
However, if performing one iteration is quick, this could have the opposite effect as excessive context switches between the function, and the event loop might slow down the execution.
To me, this function seems to fall into the second case. Thus I would suggest using a synchronous implementation.

Reply:
You are right, going with a sync `to_documents` method